vernac/indschemes.ml

```diff
@@ -321,62 +320,60 @@ let declare_sym_scheme ind =
-    ( match t with
-      | InductionScheme (x,y,z) -> ((id,x,smart_global_inductive y,z)::l1),l2
-      | CaseScheme (x,y,z) -> ((id,x,smart_global_inductive y,z)::l1),l2
-      | EqualityScheme  x -> l1,((Some id,smart_global_inductive x)::l2)
-    )
-(*
- if no name has been provided, we build one from the types of the ind
-requested
-*)
-  | (None,t)::q ->
-      let l1,l2 = split_scheme env q in
-      let names inds recs isdep y z =
-        let ind = smart_global_inductive y in
-        let sort_of_ind = inductive_sort_family (snd (lookup_mind_specif env ind)) in
-        let suffix = (
-          match sort_of_ind with
-          | InProp ->
-              if isdep then (match z with
-              | InSProp -> inds ^ "s_dep"
-              | InProp -> inds ^ "_dep"
-              | InSet  -> recs ^ "_dep"
-              | InType -> recs ^ "t_dep")
-              else ( match z with
-              | InSProp -> inds ^ "s"
-              | InProp -> inds
-              | InSet -> recs
-              | InType -> recs ^ "t" )
-          | _ ->
-              if isdep then (match z with
-              | InSProp -> inds ^ "s"
-              | InProp -> inds
-              | InSet -> recs
-              | InType -> recs ^ "t" )
-              else (match z with
-              | InSProp -> inds ^ "s_nodep"
-              | InProp -> inds ^ "_nodep"
-              | InSet -> recs ^ "_nodep"
-              | InType -> recs ^ "t_nodep")
-        ) in
-        let newid = add_suffix (Nametab.basename_of_global (GlobRef.IndRef ind)) suffix in
-        let newref = CAst.make newid in
-          ((newref,isdep,ind,z)::l1),l2
-      in
-        match t with
-        | CaseScheme (x,y,z) -> names "_case" "_case" x y z
-        | InductionScheme (x,y,z) -> names "_ind" "_rec" x y z
-        | EqualityScheme  x -> l1,((None,smart_global_inductive x)::l2)
-
-let do_mutual_induction_scheme ?(force_mutual=false) lnamedepindsort =
-  let lrecnames = List.map (fun ({CAst.v},_,_,_) -> v) lnamedepindsort
-  and env0 = Global.env() in
+  | [] -> []
+  | (Some id, {sch_type; sch_qualid; sch_sort}) :: q
+  -> ((id, sch_isdep sch_type, smart_global_inductive sch_qualid, sch_sort)
+    :: name_and_process_schemes env q)
+(* If no name has been provided, we build one from the types of the ind requested *)
+  | (None, {sch_type; sch_qualid; sch_sort}) :: q
+   -> let ind = smart_global_inductive sch_qualid in
+      let sort_of_ind = inductive_sort_family (snd (lookup_mind_specif env ind)) in
+      let suffix = scheme_suffix_gen {sch_type; sch_qualid; sch_sort} sort_of_ind in
+      let newid = add_suffix (Nametab.basename_of_global (GlobRef.IndRef ind)) suffix in
+      let newref = CAst.make newid in
+      (newref, sch_isdep sch_type, ind, sch_sort) :: name_and_process_schemes env q
+
+let do_mutual_induction_scheme ?(force_mutual=false) l =
+  let env0 = Global.env() in
+  let lnamedepindsort = name_and_process_schemes env0 l in
```

Comment:
For Equations, it would be better I think to keep a lower-level version of `do_mutual_induction_scheme`. Indeed, Equations has a kernel name, so it would be complicated to rebuild and reinterpret a qualid.

Then, assuming that `name_and_process_schemes` is called before calling `do_mutual_induction_scheme` (typically in `do_scheme`) rather than called in `do_mutual_induction_scheme`, Equations would probably go w/o a change.

Reply:
I had a go, let's see if this worked.